second homework completed

Kimislazuli opened the pull request.

Kimislazuli committed:
feat(task 1): add clients for github and stackoverflow

Kimislazuli committed:
test(task 2): test clients + wiremock-standalone instead of wiremock

Kimislazuli committed:
feat(task 3): add dummy scheduler

Kimislazuli committed:
fix: add package scrapper

Kimislazuli committed:
fix: remove push trigger from workflows

nexen505 commented on scrapper/src/main/java/edu/java/scrapper/client/GithubClient.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    public GithubClient(String baseUrl) {
+        webClient = WebClient.builder().baseUrl(baseUrl).build();
```
Проблема: тут и в соседнем клиенте при таком создании клиента вручную теряются предконфигурированные фреймворком компоненты.

Предложение: добавить при конфигурации зависимость на WebClient.Builder webClientBuilder и использовать ее для создания клиента.

nexen505 commented on scrapper/src/main/java/edu/java/scrapper/sceduler/LinkUpdaterScheduler.java:
```diff
@@ -0,0 +1,15 @@
+@Component
+public class LinkUpdaterScheduler {
+
+    @Scheduled(fixedDelayString = "#{@schedulerFixedDelay}")
```
Мысль: можно значительно упростить выражение и избавиться от лишнего бина, если воспользоваться знаком $ для получения конфигурационного свойства приложения по ключу.

Kimislazuli replied:
Да, согласна, изначально и планировала подтягивать напрямую из .yml, но немного смутила строчка про SpEL в формулировке задания `@Scheduled(fixedDelayString = "#{@beanName}")` , поэтому было решено следовать этому веянию и создать бин, хоть это и излишество своего рода. Вернула получение из файла.

А ещё в таком случае мне не очень понятна мотивация создания record Scheduler. В какой момент она используется?
```
public record ApplicationConfig(
    @NotNull
    Scheduler scheduler
) {

    public record Scheduler(boolean enable, @NotNull Duration interval, @NotNull Duration forceCheckDelay) {
    }
}
```

nexen505 replied:
Она нужна для нескольких целей:
- фреймворк через нее понимает тип конфигурируемых свойств,
- предоставление доступа к свойствам в Java через бин-компонент `@ConfigurationProperties`.

nexen505 commented on scrapper/src/main/java/edu/java/scrapper/configuration/ClientConfiguration.java:
```diff
@@ -0,0 +1,19 @@
+public class ClientConfiguration {
+    @Bean
+    public GithubClient githubClient() {
+        return new GithubClient();
```
Проблема: не выполнено требование задания по указанию базового адреса и использование значения по умолчанию, если он не указан.

Предложение: воспользоваться при конфигурации свойствами приложения через механизм `@Value`.

Kimislazuli replied:
Немного запугали в чате курса "вредностью" `@Value`, потому что она обеспечивает "слабую связность", так что я думала, что использовать её плохо (хотя контекст очевидно располагает). Хотелось бы разобаться, где находится грань уместного и неуместого использования этой аннотации.

Для решения же пункта задачи подразумевалось использование ссылки в хардкоде конструктора без аргументов (имеется также полноценный конструктор с аргументами, он используется в тестах):
```
    public GithubClient() {
        this("https://api.github.com");
    }
```

Исправлено.

Также вопрос, есть ли альтернативы в этой ситуации, что-то кроме аннотации `@Value`?

nexen505 replied:
Альтернатива: пользоваться в конфигурации бином-компонентом со свойствами, т.е. `ApplicationConfig`, и доставать из него конкретные свойства императивно, а не через ключ в аннотации.

nexen505 commented on scrapper/src/test/java/edu/java/scrapper/GithubClientTest.java:
```diff
@@ -0,0 +1,42 @@
+    @BeforeAll
+    public static void setUp() {
+        server.start();
+        stubFor(get(urlEqualTo("/repos/johndoe/java")).willReturn(aResponse()
```
Проблема: этот стаб специфичен конкретному test-case, но содержится в общей логике.

Предложение: перенести это в имеющийся `@Test`.

nexen505 commented on scrapper/src/test/java/edu/java/scrapper/StackOverflowClientTest.java:
```diff
@@ -0,0 +1,42 @@
+    @BeforeAll
+    public static void setUp() {
+        server.start();
+        server.stubFor(get(urlEqualTo("/questions/7?site=stackoverflow")).willReturn(aResponse()
```
Проблема: этот стаб специфичен конкретному test-case, но содержится в общей логике.

Предложение: перенести это в имеющийся `@Test`.

Kimislazuli commented:
Помимо прочего вызывает вопросы, в чем принципиальная разница возвращения Mono и просто упаковываемого в Mono типа...

и еще почему-то сломалось покрытие тестами.

Kimislazuli committed:
fix: post-review fixes

- add log
- use created by Spring WebClient.Builder for each client
- fetch values from application.yml instead if hardcode
- move test-specific stubs

Kimislazuli committed:
feat: change github response parsing

nexen505 commented:
> Помимо прочего вызывает вопросы, в чем принципиальная разница возвращения Mono и просто упаковываемого в Mono типа...

Принципиальная разница кроется в самом типе, собственно. `Mono` - тип, присущий реактивному программированию и стилю. Он бывает удобен в ситуации, когда весь сервис написан в реактивной парадигме и манипуляция данными происходит реактивно, так что там такие типы естественны.
